server: collect basic telemetry for our HTTP endpoints

aayushshah15 opened the pull request:
Prior to this patch, we lacked any sort of telemetry on the usage of the
HTTP endpoints exposed to users by a Cockroach cluster. This commit adds
a middleware that counts usage of the various endpoints we expose, to be
be included in the diagnostic reports we collect.

Release note: None

Release justification: Low-risk telemetry collection on the usage of
CRDB's HTTP endpoints

aayushshah15 committed:
server: remove safeServeMux

`safeServeMux` was a wrapper to provide proper locking around
`http.ServeMux` because in go1.10 and below, calls to `Handle` and
`ServeHTTP` could race with each other.

We don't need it anymore, see #29230.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/51825)
<!-- Reviewable:end -->

dhartunian commented on pkg/server/servemux_test.go:
```diff
@@ -31,10 +31,14 @@ func TestServeMuxConcurrency(t *testing.T) {
+	// fixed in go1.11.
+	// On go1.10, this test was meant for a safe wrapper that we had
+	// written to provide proper locking around an http.ServeMux.
+	var mux http.ServeMux
```
if `safeServeMux` doesn't exist anymore, I don't see why this test should be kept around. Removing sounds good to me.

aayushshah15 replied:
Done

dhartunian commented on pkg/server/server.go:
```diff
@@ -1004,6 +982,15 @@ func (s *Server) startPersistingHLCUpperBound(
+// of paths served by `inner`.
+func collectCallCounts(inner http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		telemetry.Count(fmt.Sprintf("HTTP: %s -> %s", req.Method, req.URL.EscapedPath()))
```
Can we follow the pattern of other calls to `telemetry.Count` and do something like `http.%s.%s`?

aayushshah15 replied:
Done

dhartunian commented on pkg/server/server.go:
```diff
@@ -1004,6 +982,15 @@ func (s *Server) startPersistingHLCUpperBound(
+// of paths served by `inner`.
+func collectCallCounts(inner http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		telemetry.Count(fmt.Sprintf("HTTP: %s -> %s", req.Method, req.URL.EscapedPath()))
```
I'm a bit nervous about logging URLs directly in case they contain PII or other customer data we don't want to be capturing. From a quick look at `status.proto` and `admin.proto` the URLs seem limited to containing node and range IDs but would be nice to capture the "pattern" of the URL and not the literal one being requested if we could. Not sure if that's possible.

In addition, logging URL literals could really explode the number of counters. Any concerns there? I don't know much about the storage of those and whether we need to worry about having too many keys.

aayushshah15 replied:
The `EscapedPath` of the `URL` should only include the path itself and not the query params. Here is an example of what it looks like on `./cockroach demo` after loading up the admin UI and clicking around on the metrics page a bit. 
<img width="479" alt="Screen Shot 2020-07-24 at 5 11 09 PM" src="https://user-images.githubusercontent.com/10788754/88435703-bf06b180-cdd0-11ea-912b-fdf0433e22fa.png">

Is there any way I could confirm that the URL paths themselves won't contain any PII? A cursory glance seems tells me that the list of endpoints here will be limited to the ones Alex listed in his API guidelines doc.

dhartunian replied:
I think it should be fine. I realized that since we do client-side telemetry logging as well which often exposes URLs we likely already make sure that those are "safe" to share.

@piyush-singh any thoughts on the discussion here? Also does this telemetry look like what you're expecting to capture?

dhartunian reviewed (changes requested):
Thanks for putting this up so quickly!

knz reviewed (commented):
> Reviewed 2 of 2 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51825)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dhartunian, @knz, and @piyush-singh)

___
*[pkg/server/server.go, line 989 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/51825#-MCuZbbIulKVuHBbR0jo-r2-989:-MDQRvec6SudOrwbReOu:b-l2xr3r) ([raw file](https://github.com/cockroachdb/cockroach/blob/26cd5a6f43cbf598d254db1df843db8d55e02bf1/pkg/server/server.go#L989)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

I think it should be fine. I realized that since we do client-side telemetry logging as well which often exposes URLs we likely already make sure that those are "safe" to share.

@piyush-singh any thoughts on the discussion here? Also does this telemetry look like what you're expecting to capture?
</blockquote></details>

Some API endpoints pass the argument in the path. For example database, node and table details. Can you include examples of these here? I believe database and table names are sensitive information.

I'd like to see something auto-generated from the .proto file to automatically filter out the things we don't want to (or can't ) include in telemetry.


<!-- Sent from Reviewable.io -->

knz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51825)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dhartunian, @knz, and @piyush-singh)

___
*[pkg/server/server.go, line 989 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/51825#-MCuZbbIulKVuHBbR0jo-r2-989:-MDQUEXf3TKp9XFNYL1f:bnwrqxn) ([raw file](https://github.com/cockroachdb/cockroach/blob/26cd5a6f43cbf598d254db1df843db8d55e02bf1/pkg/server/server.go#L989)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Some API endpoints pass the argument in the path. For example database, node and table details. Can you include examples of these here? I believe database and table names are sensitive information.

I'd like to see something auto-generated from the .proto file to automatically filter out the things we don't want to (or can't ) include in telemetry.
</blockquote></details>

I see the URL patterns are auto-generated as `pattern_XXX` variables in the `.gw.go` files in `serverpb`.

- We can extract the telemetry rules from that maybe?
- or (more intrusive) inject telemetry calls in the Go closures in the `.gw.go` 


<!-- Sent from Reviewable.io -->

aayushshah15 commented:
Sorry for letting this sit for the last couple weeks. I added a method that contains a blocklist of patterns (the `pattern_xxxx` variables from `.gw.go` files) that we want to ignore for the telemetry collection. 

Before: 
<img width="509" alt="Screen Shot 2020-08-13 at 2 28 27 PM" src="https://user-images.githubusercontent.com/10788754/90189331-f8f12500-dd8a-11ea-8418-aff1b74513bc.png">

After:
<img width="353" alt="Screen Shot 2020-08-13 at 5 32 00 PM" src="https://user-images.githubusercontent.com/10788754/90189342-fe4e6f80-dd8a-11ea-8ccc-ac55250e39dc.png">

Do we like this approach? If so, I'll add a few tests.

dhartunian reviewed (commented):
@aayushshah15 I like this approach and wonder if we could automate further.

Sorry I totally forgot to mention earlier but we do an automated scrubbing of sensitive fields on the CockroachCloud side through some proto extensions. If you the install a gRPC middleware (as opposed to an HTTP middleware) you can access metadata about the request/response objects that tell you if their fields should be redacted or not. I wonder if that's possible to do here.

Take a look at `scrubSensitiveFields` implementation in cloud:
https://github.com/cockroachlabs/managed-service/blob/0eaed9854375ec181878f963deaa5bc000b38dfd/pkg/auditlog/scrub.go#L14-L50

And the auditlog extension usage in the proto files:
https://github.com/cockroachlabs/managed-service/blob/4abbecde95a27f6bbdcc60827cf3281d2aca203e/console/apipb/api.proto#L226-L250

Alternatively, it might be an even simpler implementation with a gRPC middleware where you log the gRPC endpoint name instead of the URL like `DatabaseDetails` instead of `/_admin/v1/databases/{database}`.

> Reviewed 1 of 2 files at r3, 1 of 1 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51825#-:-MFHgM5R6tTMDm3G3cQY:b-izebhs)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dhartunian, @knz, and @piyush-singh)



<!-- Sent from Reviewable.io -->

aayushshah15 commented:
@dhartunian I think that gets us a much cleaner solution. I changed all this around to instead add just a grpc interceptor that logs the grpc service name (which doesn't contain any sensitive info). This is what that looks like: 
![Screenshot from 2020-08-26 11-05-43](https://user-images.githubusercontent.com/10788754/91321158-56727200-e78c-11ea-8eb3-166df3ad9fc0.png)

@knz, do you like this?

knz commented:
yes this is very adequate for th is version. Thank you.

knz commented:
can you rebase and ensure CI passes?

aayushshah15 committed:
server: collect basic telemetry for grpc-gateway HTTP endpoints

Prior to this patch, we lacked any sort of telemetry on the usage of the
HTTP endpoints exposed to users by a Cockroach cluster via grpc-gateway.
This commit adds a middleware that counts usage of the various endpoints
exposed, to be included in the diagnostic reports we collect.

Release note: None

Release justification: Low-risk telemetry collection on the usage of
CRDB's HTTP endpoints

aayushshah15 commented:
this is ready for one more look if you're interested @knz and @dhartunian

knz reviewed (commented):
I like this! but it would benefit from a unit test.

> Reviewed 1 of 1 files at r6.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51825#-:-MGD9CZa34fauNtXyMA8:bicm1rp)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aayushshah15 and @piyush-singh)

___
*[pkg/server/server.go, line 1205 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/51825#-MGD91r83EMeTKwruJhX:-MGD91r83EMeTKwruJhY:bqddart) ([raw file](https://github.com/cockroachdb/cockroach/blob/80faaed17a76328e9c1dd69d7d120923d63dc94b/pkg/server/server.go#L1205)):*
> ```Go
> 		opts ...grpc.CallOption,
> 	) error {
> 		telemetry.Count(fmt.Sprintf("http.grpc-gateway.%s", method))
> ```

can you extract that sprintf in a separate function `GetEndpointCounter` returning a `telemetry.Counter` and then `Inc` that here


<!-- Sent from Reviewable.io -->

aayushshah15 reviewed (commented):
> I like this! but it would benefit from a unit test.

done

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51825#-:-MGiIqOd4dXtZJdzfbyy:b-bfxbqj)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz and @piyush-singh)

___
*[pkg/server/server.go, line 1205 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/51825#-MGD91r83EMeTKwruJhX:-MGiIuprDwsAjMn-hUfH:b-7rl0jd) ([raw file](https://github.com/cockroachdb/cockroach/blob/80faaed17a76328e9c1dd69d7d120923d63dc94b/pkg/server/server.go#L1205)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

can you extract that sprintf in a separate function `GetEndpointCounter` returning a `telemetry.Counter` and then `Inc` that here
</blockquote></details>

done.


<!-- Sent from Reviewable.io -->

knz reviewed (approved):
thank you!

> Reviewed 1 of 2 files at r7.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51825#-:-MGiJDAY3bY3Rwhdd24d:bsv417h)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz and @piyush-singh)



<!-- Sent from Reviewable.io -->

dhartunian reviewed (approved):
LGTM. Thanks for taking some extra passes on this @aayushshah15! 
Always nice to see a PR with more line removals than additions :)

aayushshah15 commented:
TFTR! 

bors r=dhartunian,knz
